File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/sorting/quick_sort.py
def quick_sort(values: list[int]) -> list[int]:
    if len(values) <= 1:
        return values[:]

    pivot = values[len(values) // 2]
    left = [value for value in values if value < pivot]
    middle = [value for value in values if value == pivot]
    right = [value for value in values if value > pivot]
    return quick_sort(left) + middle + quick_sort(right)


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 7, 99, 1]
    print("QuickSort")
    print(f"input : {sample}")
    print(f"output: {quick_sort(sample)}")

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/sorting/QuickSort.java
package com.sortingsearching.sorting;

import java.util.Arrays;

public final class QuickSort {
    private QuickSort() {
    }

    public static int[] sort(int[] input) {
        int[] values = Arrays.copyOf(input, input.length);
        quickSort(values, 0, values.length - 1);
        return values;
    }

    private static void quickSort(int[] values, int low, int high) {
        if (low >= high) {
            return;
        }
        int pivotIndex = partition(values, low, high);
        quickSort(values, low, pivotIndex - 1);
        quickSort(values, pivotIndex + 1, high);
    }

    private static int partition(int[] values, int low, int high) {
        int pivot = values[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (values[j] <= pivot) {
                i++;
                swap(values, i, j);
            }
        }
        swap(values, i + 1, high);
        return i + 1;
    }

    private static void swap(int[] values, int i, int j) {
        int temp = values[i];
        values[i] = values[j];
        values[j] = temp;
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 7, 99, 1};
        System.out.println("QuickSort");
        System.out.println("input : " + Arrays.toString(sample));
        System.out.println("output: " + Arrays.toString(sort(sample)));
    }
}
